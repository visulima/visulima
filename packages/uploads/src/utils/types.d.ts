import Cache from "lru-cache";
import type { BinaryToTextEncoding, Hash } from "node:crypto";
import type { IncomingMessage } from "node:http";
import { Transform } from "node:stream";

export interface RangeChecksum extends Transform {
    hash: Hash;

    path: string;

    reset(): void;

    digest(encoding: BinaryToTextEncoding = "hex"): string;
}

export interface RangeHasher extends Cache<string, Hash> {
    algorithm: "sha1" | "md5";

    hex(path: string): string;

    base64(path: string): string;

    init(path: string, start = 0): Promise<Hash>;

    digester(path: string): RangeChecksum;

    updateFromFs(path: string, start = 0, initial?: Hash): Promise<Hash>;
}

export interface HttpErrorBody {
    message: string;
    code: string;
    UploadErrorCode?: string;
    name?: string;
    retryable?: boolean;
    detail?: Record<string, any> | string;
}

export interface HttpError<T = HttpErrorBody> extends UploadResponse<T> {
    statusCode: number;
}

export interface IncomingMessageWithBody<T = any> extends IncomingMessage {
    body?: T;
    _body?: boolean;
}

export type Header = number | string | string[];
export type Headers = Record<string, Header>;

export type ResponseBody = string | Record<string, any>;
export type ResponseBodyType = "text" | "json";
export type ResponseTuple<T = ResponseBody> = [statusCode: number, body?: T, headers?: Headers];

export interface UploadResponse<T = ResponseBody> extends Record<string, any> {
    statusCode?: number;
    headers?: Headers;
    body?: T;
}

export interface ValidatorConfig<T> {
    value?: any;
    isValid?: (t: T) => boolean | Promise<boolean>;
    response?: ResponseTuple<any> | HttpError<any>;
}

export type Validation<T> = Record<string, ValidatorConfig<T>>;

export interface ValidationError extends HttpError {
    name: "ValidationError";
    code: string;
}

export interface Logger {
    debug(...data: any[]): void;
    info(...data: any[]): void;
    warn(...data: any[]): void;
    error(...data: any[]): void;
}
