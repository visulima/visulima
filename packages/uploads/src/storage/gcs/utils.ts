import type { FilePart } from "../utils/file";
import { hasContent } from "../utils/file";
import GCSFile from "./gcs-file";

export function getRangeEnd(range: string): number {
    const end = +(range.split(/0-/)[1] as string);

    return end > 0 ? end + 1 : 0;
}

export function buildContentRange(part: Partial<FilePart> & GCSFile): string {
    if (hasContent(part)) {
        const end = part.contentLength ? part.start + part.contentLength - 1 : "*";
        return `bytes ${part.start}-${end}/${part.size ?? "*"}`;
    }
    return `bytes */${part.size ?? "*"}`;
}
