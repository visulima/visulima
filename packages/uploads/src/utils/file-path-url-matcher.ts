import { match } from "path-to-regexp";

const filePathUrlMatcher = match<PathMatch>(["/(.*)/:uuid.:ext", "/(.*)/:uuid"], { decode: decodeURIComponent });

export type PathMatch = { uuid: string, ext?: string };

export default filePathUrlMatcher;
