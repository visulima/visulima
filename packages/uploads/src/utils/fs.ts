import { promises as fsp } from "node:fs";
import { dirname } from "node:path";

/**
 * Ensures that the file exists and returns it size
 * @param path - filename or path to a local file
 * @param overwrite - force creating new empty file
 * @returns file size
 */
export async function ensureFile(path: string, overwrite = false): Promise<number> {
    await fsp.mkdir(dirname(path), { recursive: true });

    const open = await fsp.open(path, overwrite ? "w" : "a");
    await open.close();

    const stat = await fsp.stat(path);

    return stat.size;
}

/**
 * Removes the specified file from the local file system
 */
export async function removeFile(path: string): Promise<void> {
    if (fsp.rm) {
        return fsp.rm(path, { force: true });
    }

    return fsp.unlink(path).catch((error: NodeJS.ErrnoException): void => {
        if (error.code !== "ENOENT") throw error;
    });
}

/**
 * @internal
 */
export { promises as fsp } from "node:fs";
