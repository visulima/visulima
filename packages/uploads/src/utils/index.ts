import fnv1a from "@sindresorhus/fnv1a";
import mem from "mem";

export const hash = mem((value: string) => fnv1a.bigInt(value, { size: 64 }).toString(16));

export type { ErrorResponses } from "./errors";
export {
    ERRORS, throwErrorCode, UploadError, ErrorMap, isUploadError,
} from "./errors";
export {
    fsp, ensureFile, removeFile,
} from "./fs";
export {
    getHeader,
    appendHeader,
    setHeaders,
    extractHost,
    extractProto,
    normalizeHookResponse,
    normalizeOnErrorResponse,
    getBaseUrl,
    readBody,
    getMetadata,
    getIdFromRequest,
    uuidRegex,
    getRealPath,
} from "./http";
export { default as Locker } from "./locker";
export { default as StreamLength } from "./pipes/stream-length";
export { streamChecksum, StreamChecksum } from "./pipes/stream-checksum";
export { default as Validator, isValidationError } from "./validator";
export { default as RangeHasher } from "./range-hasher";

export { default as isEqual } from "./primitives/is-equal";
export { default as isRecord } from "./primitives/is-record";
export { default as mapValues } from "./primitives/map-values";
export { default as pick } from "./primitives/pick";
export { default as toSeconds } from "./primitives/to-seconds";
export { default as toMilliseconds } from "./primitives/to-milliseconds";

export type { PathMatch } from "./file-path-url-matcher";
export { default as filePathUrlMatcher } from "./file-path-url-matcher";

export type {
    UploadResponse,
    ResponseBody,
    ResponseTuple,
    Headers,
    Header,
    HttpErrorBody,
    HttpError,
    ResponseBodyType,
    IncomingMessageWithBody,
    Validation,
    ValidatorConfig,
    ValidationError,
    Logger,
} from "./types.d";
