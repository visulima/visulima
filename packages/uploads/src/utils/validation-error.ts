import { Headers, HttpErrorBody, ValidationError as IValidationError } from "./types";

class ValidationError extends Error implements IValidationError {
    name: "ValidationError" = "ValidationError";

    constructor(public code: string, public statusCode: number, public body: HttpErrorBody, public headers: Headers) {
        super(body?.message);
    }
}

export default ValidationError;
