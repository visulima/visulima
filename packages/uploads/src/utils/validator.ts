import { ErrorMap } from "./errors";
import type {
    Headers, HttpErrorBody, ResponseBody, ResponseTuple, UploadResponse, Validation, ValidatorConfig,
} from "./types.d";
import ValidationError from "./validation-error";

const capitalize = (s: string): string => s && (s[0] as string).toUpperCase() + s.slice(1);

const toResponse = <T extends ResponseBody>(response: ResponseTuple<T> | UploadResponse<T>): UploadResponse => {
    if (!Array.isArray(response)) {
        return response;
    }

    const [statusCode, body, headers] = response;

    return { statusCode, body, headers };
};

class Validator<T> {
    private validators: Record<string, Required<ValidatorConfig<T>>> = {};

    constructor(private prefix = "ValidationError") {}

    add(config: Validation<T>): void {
        Object.entries(config).forEach(([key, validator]) => {
            const code = `${this.prefix}${capitalize(key)}`;

            this.validators[code] = { ...this.validators[code], ...validator } as Required<ValidatorConfig<T>>;

            if (typeof (this.validators[code] as Required<ValidatorConfig<T>>).isValid !== "function") {
                throw new TypeError('Validation config "isValid" is missing, or it is not a function!');
            }
        });
    }

    async verify(t: T): Promise<void | never> {
        // eslint-disable-next-line no-restricted-syntax
        for await (const [code, validator] of Object.entries(this.validators)) {
            const isValid = await validator.isValid(t);

            if (!isValid) {
                const { statusCode, body, headers } = toResponse(validator.response || ErrorMap[code]);

                throw new ValidationError(code, statusCode as number, body as HttpErrorBody, headers as Headers);
            }
        }
    }
}

export function isValidationError(error: unknown): error is ValidationError {
    return (error as ValidationError).name === "ValidationError";
}

export default Validator;
