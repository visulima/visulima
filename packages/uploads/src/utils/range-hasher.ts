import Cache from "lru-cache";
import { Hash } from "node:crypto";
import { createReadStream } from "node:fs";

import RangeChecksum from "./range-checksum";
import type { RangeChecksum as IRangeChecksum } from "./types.d";

class RangeHasher extends Cache<string, Hash> {
    public constructor(public algorithm: "sha1" | "md5" = "sha1", options?: Cache.Options<string, Hash>) {
        super({
            ttl: 1000,
            ...options,
        });
    }

    hex(path: string): string {
        return this.get(path)?.copy().digest("hex") || "";
    }

    base64(path: string): string {
        return this.get(path)?.copy().digest("base64") || "";
    }

    async init(path: string, start = 0): Promise<Hash> {
        return this.get(path)?.copy() || this.updateFromFs(path, start);
    }

    digester(path: string): IRangeChecksum {
        return new RangeChecksum(this, path);
    }

    // eslint-disable-next-line default-param-last
    public async updateFromFs(path: string, start = 0, initial?: Hash): Promise<Hash> {
        const hash = await this.fromFs(path, start, initial);

        this.set(path, hash);

        return hash;
    }

    // eslint-disable-next-line default-param-last
    private fromFs(path: string, start = 0, initial?: Hash): Promise<Hash> {
        // eslint-disable-next-line compat/compat
        return new Promise((resolve, reject) => {
            const digester = this.digester(path);

            if (initial !== undefined) {
                digester.hash = initial;
            }

            createReadStream(path, { start })
                .on("error", reject)
                .on("end", () => resolve(digester.hash))
                .pipe(digester)
                .resume();
        });
    }
}

export default RangeHasher;
