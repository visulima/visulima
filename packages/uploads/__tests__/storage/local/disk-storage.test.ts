import {
    afterEach, beforeEach, describe, expect, it, jest,
} from "@jest/globals";
// eslint-disable-next-line unicorn/prefer-module
import { vol } from "memfs";
// eslint-disable-next-line import/no-namespace
import * as fs from "node:fs";
import { join } from "node:path";
import { createRequest } from "node-mocks-http";

import DiskStorage from "../../../src/storage/local/disk-storage";
import { fsp } from "../../../src/utils/fs";
import { metafile, storageOptions, testRoot } from "../../__helpers__/config";
import FileWriteStream from "../../__helpers__/streams/file-write-stream";
import RequestReadStream from "../../__helpers__/streams/request-read-stream";
import { deepClone } from "../../__helpers__/utils";

jest.mock("fs/promises", () => {
    // eslint-disable-next-line unicorn/prefer-module
    const process = require("node:process");
    process.chdir("/");

    // eslint-disable-next-line unicorn/prefer-module,@typescript-eslint/no-shadow
    const { fs } = require("memfs");

    // eslint-disable-next-line unicorn/prefer-module
    return {
        __esModule: true,
        ...fs.promises,
    };
});

jest.mock("fs", () => {
    // eslint-disable-next-line unicorn/prefer-module
    const process = require("node:process");
    process.chdir("/");

    // eslint-disable-next-line unicorn/prefer-module,@typescript-eslint/no-shadow
    const { fs } = require("memfs");

    // eslint-disable-next-line unicorn/prefer-module
    return {
        __esModule: true,
        ...fs,
    };
});

// eslint-disable-next-line radar/no-identical-functions
jest.mock("node:fs", () => {
    // eslint-disable-next-line unicorn/prefer-module
    const process = require("node:process");
    process.chdir("/");

    // eslint-disable-next-line unicorn/prefer-module,@typescript-eslint/no-shadow
    const { fs } = require("memfs");

    // eslint-disable-next-line unicorn/prefer-module
    return {
        __esModule: true,
        ...fs,
    };
});

const directory = join(testRoot, "disk-storage");

describe("DiskStorage", () => {
    jest.useFakeTimers({ doNotFake: ["setTimeout"] }).setSystemTime(new Date("2022-02-02"));

    const options = { ...storageOptions, directory };

    let storage: DiskStorage;
    let readStream: RequestReadStream;

    const request = createRequest();
    const createFile = (): Promise<any> => {
        storage = new DiskStorage(options);

        return storage.create(request, deepClone(metafile));
    };

    beforeEach(async () => {
        vol.reset();
    });

    afterEach(async () => {
        vol.reset();
    });

    describe("initialization", () => {
        it("should set directory", () => {
            storage = new DiskStorage(options);

            expect(storage.directory).toBe(directory);
        });
    });

    describe(".create()", () => {
        beforeEach(() => {
            storage = new DiskStorage(options);
        });

        it("should set status and bytesWritten", async () => {
            const diskFile = await storage.create(request, metafile);

            expect(diskFile).toMatchSnapshot();
        });

        it("should reject on limits", async () => {
            await expect(storage.create(request, { ...metafile, size: 6e10 })).rejects.toMatchSnapshot();
        });
    });

    describe(".update()", () => {
        beforeEach(createFile);

        it("should update metadata", async () => {
            const file = await storage.update(metafile, { metadata: { name: "newname.mp4" } });

            expect(file.metadata.name).toBe("newname.mp4");
            expect(file.metadata.mimeType).toBe("video/mp4");
        });
    });

    describe(".write()", () => {
        beforeEach(() => {
            vol.reset();

            readStream = new RequestReadStream();

            return createFile();
        });

        it("should set status and bytesWritten", async () => {
            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            const file = await storage.write({ ...metafile, start: 0, body: readStream });

            expect(file.status).toBe("part");
            expect(file.bytesWritten).toBe(5);
        });

        it("should return bytesWritten (resume)", async () => {
            const file = await storage.write({ ...metafile });

            expect(file.bytesWritten).toBe(0);
        });

        it("should reject if file not found", async () => {
            storage.cache.delete(metafile.id);

            const mockReadFile = jest.spyOn(fsp, "readFile");

            mockReadFile.mockRejectedValueOnce(new Error("not found"));

            const write = storage.write({ ...metafile });

            await expect(write).rejects.toHaveProperty("UploadErrorCode", "FileNotFound");
        });

        it("should reject on fs errors", async () => {
            const fileWriteStream = new FileWriteStream();

            jest.spyOn(fs, "createWriteStream").mockImplementationOnce(() => fileWriteStream as unknown as fs.WriteStream);

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockPipeError(fileWriteStream);

            const write = storage.write({ ...metafile, start: 0, body: readStream });
            await expect(write).rejects.toHaveProperty("UploadErrorCode", "FileError");
        });

        it("should close file and reset bytesWritten on abort", async () => {
            const fileWriteStream = new FileWriteStream();

            jest.spyOn(fs, "createWriteStream").mockImplementationOnce(() => fileWriteStream as unknown as fs.WriteStream);

            const close = jest.spyOn(fileWriteStream, "close");

            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockAbort();

            const file = await storage.write({ ...metafile, start: 0, body: readStream });

            expect(file.bytesWritten).toBeNaN();
            expect(close).toHaveBeenCalled();
        });

        it("should reject on invalid range", async () => {
            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            await expect(() => storage.write({ ...metafile, start: (metafile.size as number) - 2, body: readStream })).rejects.toThrow("File conflict");
        });

        it("should support lock", async () => {
            // eslint-disable-next-line no-underscore-dangle
            readStream.__mockSend();

            const write = storage.write({ ...metafile, start: 0, body: readStream });
            const write2 = storage.write({ ...metafile, start: 0, body: readStream });

            // eslint-disable-next-line compat/compat
            await expect(Promise.all([write, write2])).rejects.toHaveProperty("UploadErrorCode", "FileLocked");
        });
    });

    describe(".list()", () => {
        beforeEach(createFile);

        it("should return all user files", async () => {
            const items = await storage.list();

            expect(items).toHaveLength(1);
            expect(items).toMatchSnapshot();
        });
    });

    describe(".delete()", () => {
        beforeEach(createFile);

        it("should set status", async () => {
            const deleted = await storage.delete(metafile);

            expect(deleted.id).toBe(metafile.id);
            expect(deleted.status).toBe("deleted");
        });

        it("should ignore not found", async () => {
            const mockReadFile = jest.spyOn(fsp, "readFile");

            mockReadFile.mockRejectedValueOnce("notfound");

            const deleted = await storage.delete({ id: "notfound" });

            expect(deleted.id).toBe("notfound");
            expect(deleted.status).toBeUndefined();
        });
    });

    describe(".purge()", () => {
        beforeEach(createFile);

        it("should delete file", async () => {
            jest.useFakeTimers();
            jest.advanceTimersByTime(500);

            const list = await storage.purge(5);

            expect(list.items).toHaveLength(1);

            jest.useRealTimers();
        });
    });

    describe(".copy()", () => {
        beforeEach(createFile);

        it("should copy file", async () => {
            await storage.copy(`${metafile.id}.META`, `${directory}/newname1.META`);

            const list = await storage.list();

            expect(list).toHaveLength(2);
        });
    });

    describe(".move()", () => {
        beforeEach(createFile);

        it("should move file", async () => {
            await storage.move(`${metafile.id}.META`, `${directory}/newname2.META`);

            const list = await storage.list();

            expect(list).toHaveLength(1);
        });
    });
});
