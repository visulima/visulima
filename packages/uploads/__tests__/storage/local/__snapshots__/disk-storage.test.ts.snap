// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiskStorage .create() should reject on limits 1`] = `[ValidationError]`;

exports[`DiskStorage .create() should set status and bytesWritten 1`] = `
File {
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
}
`;

exports[`DiskStorage .list() should return all user files 1`] = `
[
  {
    "createdAt": 2022-02-02T00:00:00.000Z,
    "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
    "modifiedAt": 2022-02-02T00:00:00.000Z,
  },
]
`;
